src/radical/pilot/agent/scheduler/continuous.py

```diff
@@ -70,258 +76,387 @@ def finalize_child(self):
-        all_slots_slot_index = self.slots.index(slot_entry)
+    def _release_slot(self, slots):
 
-        return all_slots_slot_index * self._lrms_cores_per_node + int(first_slot_core)
+        self._change_slot_states(slots, rpc.FREE)
 
 
     # --------------------------------------------------------------------------
     #
-    def _release_slot(self, opaque_slots):
+    def _alloc_node(self, node, requested_cores, requested_gpus, 
+                    chunk=1, partial=False):
+        '''
+        Find up to the requested number of free cores and gpus in the node.
+        This call will return two lists, for each matched set.  If the core does
+        not have sufficient free resources to fullfill *both* requests, two
+        empty lists are returned.  The call will *not* change the allocation
+        status of the node, atomicity must be guaranteed by the caller.
+
+        We don't care about continuity within a single node - cores `[1,5]` are
+        assumed to be as close together as cores `[1,2]`.
+
+        When `chunk` is set, only sets of exactly that size (or multiples
+        thereof) are considered valid allocations.  The use case is OpenMP
+        support, where each process is expected to create a certain number of
+        threads, which thus must have slots on the same node available.
+
+        NOTE: chunking is only applied to cores at this point.
+
+        When `partial` is set to `True`, we also accept less cores and gpus then
+        requested (but the call will never return more than requested).
+        '''
+
+
+        cores = list()
+        gpus  = list()
+
+        free_cores = node['cores'].count(rpc.FREE)
+        free_gpus  = node['gpus' ].count(rpc.FREE)
+
+        # first count the free cores/gpus, as that is way quicker than
+        # actually finding the core IDs.
+        if partial:
+            # For partial requests the check simpliefies: we just check if we 
+            # have either, some cores *or* gpus, to serve the request
+            if  (requested_cores and not free_cores) and \
+                (requested_gpus  and not free_gpus )     :
+                # wa can't serve either request
+                return [], []
 
-        if not 'task_slots' in opaque_slots:
-            raise RuntimeError('insufficient information to release slots via %s: %s' \
-                    % (self.uid, opaque_slots))
+        else:
+            # non-partial requests (ie. full requests): check if we can serve
+            # both, requested cores *and* gpus.
+            if  requested_cores > free_cores or \
+                requested_gpus  > free_gpus     :
+                return [], []
 
-        self._change_slot_states(opaque_slots['task_slots'], rpc.FREE)
 
+        # we can serve the partial or full request - alloc the chunks we need
+        # FIXME: chunk gpus, too?
+        alloc_cores = min(requested_cores, free_cores) / chunk * chunk
```

Comment:
I think parenthesis for securing the division, multiplication order makes the code safer.